src/coloring/backtracking_coloring.jl

```diff
@@ -0,0 +1,238 @@
+                deleteat!(U, 1)
+                freeColors[x] = copy(U)
+                if i==1
+                    l = 0
+                else
+                    l = colors[i-1]
+                end
+                colors[i] = max(k, l)
+            else
+                start = i-1
+                back = true
+                break
+            end
+        end
+
+        if back
+            if start >= 1
+                x = A[start]
+                F[x] = 0
+                U = freeColors[x]
+            end
+        else
+            Fopt = copy(F)
+            opt = colors[v-1]
+            i = least_index(F,A,opt)
+            start = i-1
+            if start < 1
+                break
+            end
+
+            #uncolor all vertices A[i] with i >= start
+            uncolor_all!(F, A, start)
+
+            for i = 1:start+1
+                x = A[i]
+                U = freeColors[x]
+
+                #remove colors >= opt from U
+                U = remove_higher_colors(U, opt)
+                freeColors[x] = copy(U)
+            end
+        end
+
+    end
+
+    Fopt
+
+end
+
+"""
+    sort_by_degree(g::LightGraphs.AbstractGraph)
+
+Returns a list of the vertices of graph g sorted
+in non-increasing order of their degrees
+"""
+function sort_by_degree(g::LightGraphs.AbstractGraph)
+    vs = vertices(g)
+    degrees = (LightGraphs.degree(g, v) for v in vs)
+    vertex_pairs = collect(zip(vs, degrees))
+    sort!(vertex_pairs, by = p -> p[2], rev = true)
+    [v[1] for v in vertex_pairs]
+end
+
+"""
+    uncolored_vertex_of_maximal_degree(A::Array{Int,1},F::Array{Int,1})
+
+Returns an uncolored vertex from the partially
+colored graph which has the highest degree
+"""
+function uncolored_vertex_of_maximal_degree(A::Array{Int,1},F::Array{Int,1})
+    for v in A
+        if F[v] == 0
+            return v
+        end
+    end
+end
+
+
+"""
+    free_colors(x::Int,
+                A::Array{Int,1},
+                colors::Array{Int,1},
+                F::Array{Int64,1},
+                g::LightGraphs.AbstractGraph,
+                opt::Int)
+
+Returns set of free colors of x which are less
+than optimal color number (opt)
+
+Arguments:
+
+x: Vertex who's set of free colors is to be calculated
+A: List of vertices of graph g sorted in non-increasing order of degree
+colors: colors[i] stores the number of distinct colors used in the
+        coloring of vertices A[0], A[1]... A[i-1]
+F: F[i] stores the color of vertex i
+g: Graph to be colored
+opt: Current optimal number of colors to be used in the coloring of graph g
+"""
+function free_colors(x::Int,
+                    A::Array{Int,1},
+                    colors::Array{Int,1},
+                    F::Array{Int64,1},
+                    g::LightGraphs.AbstractGraph,
+                    opt::Int)
+    index = -1
+
+    freecolors = zeros(Int, 0)
+
+    for i in eachindex(A)
+        if A[i] == x
+            index = i
+            break
+        end
+    end
```

Comment:
is it possible that `index=-1` after the loop?

Reply:
No. Since `x` is a vertex in the graph, we can be sure that it'll be present in vector `A` as well. So index would always lie between `[1, nv(g)]` once that code block executes.